File: eventree/eventree.py

__author__ = 'jared'

from sys import stdin, stdout

"""
Problem:
You are given a tree (a simple connected graph with no cycles). You have to remove
as many edges from the tree as possible to obtain a forest with the condition that : 
Each connected component of the forest should contain an even number of vertices.

Link: https://www.hackerrank.com/challenges/even-tree
"""


class TreeGraph():

    adj_list = {}
    visited = {}

    def add_edge(self, node_1, node_2):

        if not node_1 in self.adj_list:
            self.adj_list[node_1] = set()
        self.adj_list[node_1].add(node_2)

        if not node_2 in self.adj_list:
            self.adj_list[node_2] = set()
        self.adj_list[node_2].add(node_1)

    def decompose(self):
        cuts = 0
        for node in self.adj_list:
            for edge_node in self.adj_list[node].copy():
                if (self.__number_connected_is_even(node, edge_node)
                    and self.__number_connected_is_even(edge_node, node)):
                    self.__cut(node, edge_node)
                    cuts += 1
        return cuts

    def __cut(self, node1, node2):
        self.adj_list[node1].remove(node2)
        self.adj_list[node2].remove(node1)

    def __nodes_in_subtree(self, the_node, skip_node=None):
        self.visited = { the_node: True }
        if skip_node is not None:
            self.visited[skip_node] = True
        return self.__number_connected(the_node) + 1

    def __number_connected(self, the_node):
        num_connected = 0
        for node in self.adj_list[the_node]:
            if not self.visited.get(node):
                self.visited[node] = True
                num_connected += 1 + self.__number_connected(node)
        return num_connected

    def __number_connected_is_even(self, the_node, skip_node=None):
        return self.__nodes_in_subtree(the_node, skip_node) % 2 == 0

    def __degree(self, node):
        return len(self.adj_list[node])


def read_graph():

    num_vertices, num_edges = stdin.readline().split()
    num_vertices, num_edges = int(num_vertices), int(num_edges)
    graph = TreeGraph()

    for i in range(num_edges):
        node_1, node_2 = stdin.readline().split()
        node_1, node_2 = int(node_1), int(node_2)
        graph.add_edge(node_1, node_2)

    return graph

if __name__ == "__main__":

    graph = read_graph()
    num_cuts = graph.decompose()
    print(num_cuts)

File: algorithms/graph-theory/eventree/eventree.php
<?php
/**
 * @author Jared
 */

/**
 * Problem:
 * You are given a tree (a simple connected graph with no cycles). You have to remove
 * as many edges from the tree as possible to obtain a forest with the condition that : 
 * Each connected component of the forest should contain an even number of vertices.
 *
 * See full problem: https://www.hackerrank.com/challenges/even-tree
 */


class TreeGraph {

    /** @var array the adjacency map */
    private $_adjList = array();

    /** @var array the visited map */
    private $_visited = array();

    /** decompose the Graph */
    public function decompose() {
        $cuts = 0;
        foreach (array_keys($this->_adjList) as $node) {
            foreach ($this->_adjList[$node] as $edgeNode => $value) {
                if ($this->_nodesInSubtreeIsEven($node, $edgeNode)
                    && $this->_nodesInSubtreeIsEven($edgeNode, $node)) {
                    $this->_cut($node, $edgeNode);
                    $cuts += 1;
                }
            }
        }
        return $cuts;
    }

    /**
     * Add an edge to the graph
     * @param integer $node1 a node
     * @param integer $node2 a node
     */
    public function addEdge($node1, $node2) {

        $this->_addEdgeNode($node1, $node2);
        $this->_addEdgeNode($node2, $node1);
    }

    /**
     * Add an edge to the graph
     * @param integer $mainNode adjacency node
     * @param integer $adjacentNode connected node
     */
    private function _addEdgeNode($mainNode, $adjacentNode) {

        if (!isset($this->_adjList[$mainNode])) {
            $this->_adjList[$mainNode] = array();
        }
        $this->_adjList[$mainNode][$adjacentNode] = true;
    }

    /**
     * Cut an edge on the graph
     * @param integer $node1 a node
     * @param integer $node2 a node
     */
    private function _cut($node1, $node2) {
        unset($this->_adjList[$node1][$node2]);
        unset($this->_adjList[$node2][$node1]);
    }

    /**
     * Is the number of nodes in the subtree even?
     * @param integer $theNode node to start counting from
     * @param integer $skipNode adjecent nodes to skip
     * @return boolean
     */
    private function _nodesInSubtreeIsEven($theNode, $skipNode=null) {
        $this->_visited = array($theNode => true);
        if ($skipNode !== null) {
            $this->_visited[$skipNode] = true;
        }
        return ($this->_numberConnected($theNode) + 1) % 2 == 0;
    }

    /**
     * Number connected to the node
     * @param integer $theNode node to start counting from
     * @return integer
     */
    private function _numberConnected($theNode) {
        $num_connected = 0;
        foreach ($this->_adjList[$theNode] as $node => $value) {
            
            if (!isset($this->_visited[$node])) {
                $this->_visited[$node] = true;
                $num_connected += $this->_numberConnected($node) + 1;
            }
        }
        return $num_connected;
    }
}

/**
 * Read the graph from Stdin
 * @return TreeGraph
 */
function readGraph() {

    list($num_vertices, $num_edges) = explode(" ", trim(fgets(STDIN)));
    $num_vertices = (int) $num_vertices;
    $num_edges = (int) $num_edges;
    $graph = new TreeGraph();

    for ($i = 0; $i < $num_edges; $i++) {
        list($node1, $node2) = explode(" ", trim(fgets(STDIN)));
        $node1 = (int) $node1;
        $node2 = (int) $node2;
        $graph->addEdge($node1, $node2);
    }
    return $graph;
}

    $graph = readGraph();
    $num_cuts = $graph->decompose();
    echo $num_cuts;
